src/wp-admin/includes/class-rollback-auto-update.php

```diff
@@ -0,0 +1,405 @@
+		$current_auto_updates = array_intersect( $auto_updates, $current_plugins );
+
+		// Get array of non-fatal auto-updates remaining.
+		$remaining_auto_updates = array_diff( $current_auto_updates, self::$processed, self::$fatals );
+
+		return $remaining_auto_updates;
+	}
+
+	/**
+	 * Sends an email noting successful and failed updates.
+	 */
+	private function send_update_result_email() {
+		add_filter( 'auto_plugin_theme_update_email', array( $this, 'props' ), 10, 4 );
+		$successful = array();
+		$failed     = array();
+
+		/*
+		 * Using `get_plugin_data()` instead has produced warnings/errors
+		 * as the files may not be in place at this time.
+		 */
+		$plugins = get_plugins();
+
+		foreach ( static::$current->response as $k => $update ) {
+			$item = static::$current->response[ $k ];
+			$name = $plugins[ $update->plugin ]['Name'];
+
+			/*
+			 * This appears to be the only way to get a plugin's older version
+			 * at this stage of an auto-update when not implementing this
+			 * feature directly in Core.
+			 */
+			$current_version = static::$current->checked[ $update->plugin ];
+
+			/*
+			 * The `current_version` property does not exist yet. Add it.
+			 *
+			 * `static::$current->response[ $k ]` is an instance of `stdClass`,
+			 * so this should not fall victim to PHP 8.2's deprecation of
+			 * dynamic properties.
+			 */
+			$item->current_version = $current_version;
+
+			$plugin_result = (object) array(
+				'name' => $name,
+				'item' => $item,
+			);
+
+			$success = array_diff( self::$processed, self::$fatals );
+
+			if ( in_array( $update->plugin, $success, true ) ) {
+				$successful['plugin'][] = $plugin_result;
+				continue;
+			}
+
+			if ( in_array( $update->plugin, self::$fatals, true ) ) {
+				$failed['plugin'][] = $plugin_result;
+			}
+		}
+
+		$automatic_upgrader      = new WP_Automatic_Updater();
+		$send_plugin_theme_email = new \ReflectionMethod( $automatic_upgrader, 'send_plugin_theme_email' );
+		$send_plugin_theme_email->setAccessible( true );
+		$send_plugin_theme_email->invoke( $automatic_upgrader, 'mixed', $successful, $failed );
+	}
+
+	/**
+	 * Credit where credit is due.
+	 *
+	 * @param array  $email {
+	 *     Array of email arguments that will be passed to wp_mail().
+	 *
+	 *     @type string $to      The email recipient. An array of emails
+	 *                           can be returned, as handled by wp_mail().
+	 *     @type string $subject The email's subject.
+	 *     @type string $body    The email message body.
+	 *     @type string $headers Any email headers, defaults to no headers.
+	 * }
+	 * @param string $type               The type of email being sent. Can be one of 'success', 'fail', 'mixed'.
+	 * @param array  $successful_updates A list of updates that succeeded.
+	 * @param array  $failed_updates     A list of updates that failed.
+	 *
+	 * @return array
+	 */
+	public function props( $email, $type, $successful_updates, $failed_updates ) {
```

Comment:
Method might need an update/removal for a Core merge

Reply:
Why?

Reply:
Sorry, I should have elaborated. Since this method adds the failure message, it's not really "props", or "credit where credit is due". It's more "Adds a message to note that failures occurred and rollbacks may have been performed."